File: src/lib/rubeno/lib/base_when.rb
module Rubeno
  class BaseWhen < BaseAction
    attr_accessor :adapter
    
    def initialize(name, when_cb)
      super(name, when_cb)
      @adapter = nil
    end
    
    def perform_action(store, action_cb, test_resource, artifactory = nil)
      # Default implementation that can be overridden
      # This should call the adapter's execute method
      if @adapter && @adapter.respond_to?(:execute)
        @adapter.execute(store, action_cb, test_resource, artifactory)
      else
        # Fallback to calling the action_cb directly
        if action_cb.respond_to?(:call)
          action_cb.call(store)
        else
          store
        end
      end
    end
    
    # Alias for BDD pattern - kept for compatibility but uses universal method internally
    def and_when(store, when_cb, test_resource_configuration, artifactory = nil)
      perform_action(store, when_cb, test_resource_configuration, artifactory)
    end
  end
end

File: src/lib/pitono/src/base_when.py
from typing import Any, Dict, List, Callable, Optional
from .base_action import BaseAction

class BaseWhen(BaseAction):
    def __init__(self, name: str, when_cb: Callable[[Any], Any]):
        super().__init__(name, when_cb)
    
    async def and_when(
        self,
        store: Any,
        when_cb: Callable[[Any], Any],
        test_resource_configuration: Any,
        artifactory: Any = None
    ) -> Any:
        # This should be implemented by subclasses
        # But provide a default implementation that calls perform_action
        return await self.perform_action(store, when_cb, test_resource_configuration, artifactory)
    
    async def perform_action(
        self,
        store: Any,
        action_cb: Callable[[Any], Any],
        test_resource,
        artifactory: Any = None
    ) -> Any:
        # Default implementation that can be overridden
        if callable(action_cb):
            return action_cb(store)
        return store
